### PR TITLE
Introduce ELO rating system

### PR DESCRIPTION
**Pull request by visitorckw**
Implement the ELO rating system [1] to evaluate the relative strengths of various agents based on their performance. The ELO rating system, commonly utilized in assessing the skill levels of chess players, offers a reliable method to gauge the competitive abilities of agents in our context. This addition enhances our ability to make informed decisions and comparisons within our system.

Additionally, zobrist_destroy_table() function was added to free all allocated memory associated with the hash table to ensure the proper functioning of the ELO rating system.

Link: https://en.wikipedia.org/wiki/Elo_rating_system [1]

### CODE REVIEW

**Comment by ndsl7109256** [comment]
- When RL agent play a move, it would print the Q value for available moves. I thought these info can be removed when evaluating ELO.
- Some guidance could be added to the README, and a build check could be included for GitHub Actions.

File: elo.c [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+static void dump_elo()
+{
+    printf("%-10s | %-10s | %-10s | %-10s | %-10s\n", "Agent Name",
+           "Elo Rating", "Win", "Draw", "Lose");
+    printf("---------------------------------------------------------\n");
+
+    for (int i = 0; i < 3; ++i)
+        printf("%-10s | %-11.2f | %-10lld | %-10lld | %-10lld\n", agent_name[i],
```
Comment (ndsl7109256) [inline elo.c]: Typo. `%-11.2f` should be `%-10.2f`

**Comment by visitorckw** [comment]
> * When RL agent play a move, it would print the Q value for available moves. I thought these info can be removed when evaluating ELO.

I'm thinking maybe we should default to turning off the related output messages unless the output option is enabled during compilation?

**Comment by ndsl7109256** [comment]
> I'm thinking maybe we should default to turning off the related output messages unless the output option is enabled during compilation?

It sounds good, these messages are intended for debug purposes and will not need to be displayed  most of the time.

File: Makefile [hunk 2]
```diff
@@ -8,6 +8,9 @@ MCTS = mcts
 RL_CFLAGS := $(CFLAGS) -D USE_RL
 MCTS_CFLAGS := $(CFLAGS) -D USE_MCTS
 MCTS_LDFLAGS := $(LDFLAGS) -lm
+ELO = elo
+ELO_CFLAGS := $(CFLAGS) -lm
```
Comment (jserv) [inline Makefile]: Drop `-lm` from `CFLAGS`.

File: elo.c [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+#include "game.h"
+#include "zobrist.h"
+
+#define N_GAMES 100
+#define ELO_INIT 1500
+#define ELO_K 32
+
+const char *agent_name[] = {"Negamax", "MCTS", "RL"};
+
+long long int win[] = {0, 0, 0};
```
Comment (jserv) [inline elo.c]: Can you declare these variables as `static`?

**Comment by visitorckw** [comment]
> > I'm thinking maybe we should default to turning off the related output messages unless the output option is enabled during compilation?
> 
> It sounds good, these messages are intended for debug purposes and will not need to be displayed most of the time.


Since the RL agent logging messages is not directly related to introducing the ELO system, I will file another PR to address this.

**Review by ndsl7109256 — APPROVED** [review]

**Comment by jserv** [comment]
Thank @visitorckw for contributing!

**Pull request merged**